planner/cascades/transformation_rules.go

```diff
@@ -710,6 +711,55 @@ func (r *PushLimitDownProjection) OnTransform(old *memo.ExprIter) (newExprs []*m
 	return []*memo.GroupExpr{projExpr}, true, false, nil
 }
 
+// PushLimitDownUnionAll pushes limit to union all.
+type PushLimitDownUnionAll struct {
+	baseRule
+}
+
+// NewRulePushLimitDownUnionAll creates a new Transformation PushLimitDownUnionAll.
+// The pattern of this rule is `Limit->UnionAll->X`.
+func NewRulePushLimitDownUnionAll() Transformation {
+	rule := &PushLimitDownUnionAll{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandLimit,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandUnionAll, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+// Use appliedRuleSet in GroupExpr to avoid re-apply rules.
+func (r *PushLimitDownUnionAll) Match(expr *memo.ExprIter) bool {
+	return !expr.GetExpr().HasAppliedRule(r)
+}
+
+// OnTransform implements Transformation interface.
+// It will transform `Limit->UnionAll->X` to `Limit->UnionAll->Limit->X`.
+func (r *PushLimitDownUnionAll) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	limit := old.GetExpr().ExprNode.(*plannercore.LogicalLimit)
+	unionAll := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalUnionAll)
+
+	newLimit := plannercore.LogicalLimit{
+		Count: limit.Count + limit.Offset,
+	}.Init(limit.SCtx(), limit.SelectBlockOffset())
+
+	newUnionAllExpr := memo.NewGroupExpr(unionAll)
+	for _, childGroup := range old.Children[0].GetExpr().Children {
+		newLimitExpr := memo.NewGroupExpr(newLimit)
+		newLimitExpr.Children = append(newLimitExpr.Children, childGroup)
+		newLimitGroup := memo.NewGroupWithSchema(newLimitExpr, childGroup.Prop.Schema)
+
+		newUnionAllExpr.Children = append(newUnionAllExpr.Children, newLimitGroup)
+	}
+
+	newLimitExpr := memo.NewGroupExpr(limit)
+	newUnionAllGroup := memo.NewGroupWithSchema(newUnionAllExpr, unionAll.Schema())
```

Comment:
`LogicalPlan.Schema()` cannot be used in cascades planner anymore. Use `GroupExpr.Schema()` instead.

Reply:
Fixed. Could you explain a little bit why use GroupExpr.Schema() instead.

Reply:
You can see the implementation of `LogicalPlan.Schema()`. There are two kinds of implementation, one is in `baseLogicalPlan`:
```golang
func (p *baseLogicalPlan) Schema() *expression.Schema {
	return p.children[0].Schema()
}
```
And the other one is in `logicalSchemaProducer`:
```golang
func (s *logicalSchemaProducer) Schema() *expression.Schema {
	if s.schema == nil {
		s.schema = expression.NewSchema()
	}
	return s.schema
}
```
Not all of the LogicalPlans have their own schema. For example, LogicalSort only change the order of rows without changing its schema. So LogicalSort's schema should be the same with its child's. But as for LogicalProjection or LogicalAggregation, they have their own schema instead of using their child's (so we call them schemaProducer).

But in cascades optimizer, we don't use the schema of LogicalPlan anymore. You can see that you created a new `LogicalLimit`, but you didn't set it's children(you only set GroupExpr's children). So if you want to get `newLimit.Schema()`, it will panic(it will call newLimit.children[0].Schema(), but newLimit.children is empty). The main reason is that, we save the schema in `Group.Prop` instead of `logicalSchemaProducer`.